pkg/apis/wksprovider/controller/wksctl/machine_controller.go

```diff
@@ -91,34 +96,92 @@ type FootlooseMachine struct {
-	contextLog := log.WithFields(log.Fields{"context": ctx, "cluster": *cluster, "machine": *machine})
-	contextLog.Info("creating machine...")
-	if err := a.create(ctx, cluster, machine); err != nil {
-		contextLog.Errorf("failed to create machine: %v", err)
-		return err
+func (r *MachineController) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.TODO() // upstream will add this eventually
+	contextLog := log.WithField("name", req.NamespacedName)
+
+	// request only contains the name of the object, so fetch it from the api-server
+	bmm := &baremetalspecv1.BareMetalMachine{}
+	err := r.client.Get(ctx, req.NamespacedName, bmm)
+	if err != nil {
+		if apierrs.IsNotFound(err) { // isn't there; give in
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
 	}
-	return nil
-}
 
-func (a *MachineActuator) create(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
-	c, m, err := a.parse(cluster, machine)
+	// Get Machine via OwnerReferences
+	machine, err := util.GetOwnerMachine(ctx, r.client, bmm.ObjectMeta)
 	if err != nil {
-		return err
+		return ctrl.Result{}, err
 	}
-	installer, closer, err := a.connectTo(machine, c, m)
+	if machine == nil {
+		contextLog.Info("Machine Controller has not yet set ownerReferences")
+		return ctrl.Result{}, nil
+	}
+	contextLog = contextLog.WithField("machine", machine.Name)
+
+	// Get Cluster via label "cluster.x-k8s.io/cluster-name"
+	cluster, err := util.GetClusterFromMetadata(ctx, r.client, machine.ObjectMeta)
+	if err != nil {
+		contextLog.Info("Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, nil
+	}
+
+	if util.IsPaused(cluster, bmm) {
+		contextLog.Info("BareMetalMachine or linked Cluster is marked as paused. Won't reconcile")
+		return ctrl.Result{}, nil
+	}
+	contextLog = contextLog.WithField("cluster", cluster.Name)
+
+	// Now go from the Cluster to the BareMetalCluster
+	bmc := &baremetalspecv1.BareMetalCluster{}
+	if err := r.client.Get(ctx, client.ObjectKey{
+		Namespace: bmm.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}, bmc); err != nil {
+		contextLog.Info("BareMetalCluster is not available yet")
+		return ctrl.Result{}, nil
+	}
+
+	// Object still there but with deletion timestamp => run our finalizer
+	if !bmm.ObjectMeta.DeletionTimestamp.IsZero() {
+		err := r.delete(ctx, bmc, machine, bmm)
+		if err != nil {
+			contextLog.Errorf("failed to delete machine: %v", err)
+		}
+		return ctrl.Result{}, err
+	}
+
+	// FIXME!  assuming everything else is create
+	{
+		err := r.create(ctx, bmc, machine, bmm)
+		if err != nil {
+			contextLog.Errorf("failed to create machine: %v", err)
+		}
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (a *MachineController) create(ctx context.Context, c *baremetalspecv1.BareMetalCluster, machine *clusterv1.Machine, bmm *baremetalspecv1.BareMetalMachine) error {
```

Comment:
Could this be called `reconcileNormal` or `reconcile` instead? This will be consistent with other providers.